projects/hathor-wallet-mobile/0003-token-swap.md

```diff
@@ -0,0 +1,98 @@
+
+Failure to resolve the list of tokens will block the swap screen from fully loading.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Token Swap UI
+
+The token swap button on the main screen will be loaded depending on a feature flag, once clicked the allowed token list will be fetched during an initial load step before the user can interact with the swap form.
+
+The screen should track if the user last changed the input token or the output token, this information is important to determine which method to call from the contract and how the swap will be made.
+
+### Choosing tokens for the swap
+
+Once the user selects both tokens and writes a value on either input or output token area the wallet will start a cancellable async task to fetch the swap information from the contract.
+
+The fetching of the swap information has to be a cancellable async task because the user may change the tokens during the request so we will need to stop the current request and start a second one with the new tokens chosen.
+
+When the exchange rate is known we will be able to calculate the desired values for the swap.
+These values will be locked in when the user clicks the review button.
+
+## Dozer token swap protocol methods
+
+### User defined the input tokens
+
+We will call the following method to receive the swap information.
+
+`find_best_swap_path(self, amount_in: Amount, token_in: TokenUid, token_out: TokenUid, max_hops: int)`
+- `amount_in`: The amount of input tokens
+- `token_in`: The input token uid
+- `token_out`: The output token uid
+- `max_hops`: Maximum number of hops (1-3, but algorithm handles any number)
+
+The path will define the next method to call.
+if only 1 swap is detected on the path `swap_exact_tokens_for_tokens(self, ctx: Context, fee: Amount)` should be called, the input and output token and amounts should be defined in the actions.
+
+If more than 1 swap is detected on the path `swap_exact_tokens_for_tokens_through_path(self, ctx: Context, path_str: str)` should be called, the input and output token and amounts should be defined in the actions.
+
+### User defined the output tokens
+
+We will call the following method to receive the swap information.
+
+`find_best_swap_path_exact_output(self, amount_in: Amount, token_in: TokenUid, token_out: TokenUid, max_hops: int)`
+- `amount_in`: The amount of input tokens
+- `token_in`: The input token uid
+- `token_out`: The output token uid
+- `max_hops`: Maximum number of hops (1-3, but algorithm handles any number)
+
+The path will define the next method to call.
+if only 1 swap is detected on the path `swap_tokens_for_exact_tokens(self, ctx: Context, fee: Amount)` should be called, the input and output token and amounts should be defined in the actions.

```

Comment:
question: Again, it's not clear to me this issue of outstanding balances. The `exact_tokens` in the name of this method imply that the other `find_best_swap_path` does not necessarily use the exact numbers when making its calculations, and have some margin of error to calculate a larger or lessesr number of tokens.

If this is the case, I think this should be made clear to the user.